### ISSUE TITLE
Azure support for uploader

### ISSUE THREAD

**Issue by schaefi** [SUSE/mash#197]
So far the uploader backend supports EC2 but should also support Azure

* https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python#mgmt-auth-file

**Comment by schaefi** [comment]
Started to work on this and was thinking if the creation of the resource group, storage account and container is also something we need to address in the uploader code or can we expect that to exist as kind of a pre-setup done for the azure account used ?

**Comment by schaefi** [comment]
For my tests I pre setup the following:

- resource group ms_group
- storage account ms1storage in ms_group resource
- storage container ms1container in ms1storage account

If mash uploader service creates an image from the vhd page blob uploaded to ms1container it uses the sdk method compute_client.images.create_or_update() which requires among other a blob_uri

```python
'blob_uri': 'https://{0}.blob.core.windows.net/{1}/{2}'.format(
    'ms1storage', 'ms1container', 'ms_test_blob'
)
```

@rjschwei @bear454: If it is possible to delete the blob uri after an image has been created from it we could design the uploader to work with temporary resource_groups only present while mash is uploading. However I don't know if that is possible and before I start to test that you might have already done this and maybe could share your experience ?

Thanks

**Comment by schaefi** [comment]
Ah and last but not least, if we consider temporary resource groups I found that the creation of the resource group, the storage account and the container takes some time. We would surely have a performance win if they would be generated only once and re-used in the context of mash. But that goes into the same direction as with the ssh keys in ec2...

**Comment by smarlowucf** [comment]
> @smarlowucf am I thinking in the right direction ? and you plan to update the credentials service for azure authentication ?

Well all credentials will be store as encrypted json strings. For both GCE and Azure the authentication is by passing a path to the json file.

The way I chose to go with Testing service was the credentials services stays the same. It passes the encrypted json and this is decrypted by each service for each set of credentials.

When an Azure test is initiated a temp file is created and the json is dumped to that file.

https://github.com/SUSE/mash/pull/201/files#diff-64cfadab481ef9cfdd38710ded6af679R53

**Comment by smarlowucf** [comment]
> Ah and last but not least, if we consider temporary resource groups I found that the creation of the resource group, the storage account and the container takes some time. We would surely have a performance win if they would be generated only once and re-used in the context of mash. But that goes into the same direction as with the ssh keys in ec2...

I think they would need to be configured for Azure uploading right? Because if we have a full publishing job then the images should go into the correct storage account/container etc?

**Comment by schaefi** [comment]
> I think they would need to be configured for Azure uploading right? 

yes that's what I thought too, so if we expect the storage account and the container in it to be present when uploading that information should be stored, probably in the job description. For ec2 the target is specified in ```target_regions``` which needs the region an account name and the helper ami. For azure ```target_regions``` would need the region a resource group a storage account name and a container name. So this ends up in a change of the job document for Azure did you and Robert had a different plan in mind ?

**Comment by schaefi** [comment]
> Well all credentials will be store as encrypted json strings. 

Well and that's basically my question. Before asking I have read services/credentials/service.py and
there the method _retrieve_credentials(). As you said the return from _get_credentials_response()
will be an encrypted json and that's clear to me. But looking at _retrieve_credentials() I see below a
TODO a special handling for the testing service which I consider to go away and all the rest is the handling for ec2 credentials. I could not find code which handles Azure. That's why I asked if you will make the credentials service aware of it. For Azure the following information needs to be encrypted:
```
{
  "clientId": "...",
  "clientSecret": "...",
  "subscriptionId": "...",
  "tenantId": "...",
  "activeDirectoryEndpointUrl": "https://login.microsoftonline.com",
  "resourceManagerEndpointUrl": "https://management.azure.com/",
  "activeDirectoryGraphResourceId": "https://graph.windows.net/",
  "sqlManagementEndpointUrl": "https://management.core.windows.net:8443/",
  "galleryEndpointUrl": "https://gallery.azure.com/",
  "managementEndpointUrl": "https://management.core.windows.net/"
}
```

If I can rely on that response from the credentials service I would write that information as NamedTemporaryFile and use it with the get_client_from_auth_file() in the Azure sdk which really
works well

Hope that clarifies the background of my question

PS:
while reading the credentials code I still see that the abstraction for credentials of the different csp's like I had planned it before you took over the service is still there but unused. So there is still an unused amazon.py which holds the record layout for ec2 auth data and I see that to be duplicated in service.py. I think it would be good and easier to read if there would be an interface. But that's just something I stumbled over and maybe you and Robert also had a different idea of how that should be done ?

**Comment by smarlowucf** [comment]
> So I think this ends up in a change of the job document for Azure

Yes, I think there will be differences in job doc and accounts files. I.e. for Azure I don't think we need the list of regions per partition. It looks like the regions list can be retrieved based on the account even for sovereign regions which is not the case directly for EC2.

**Comment by smarlowucf** [comment]
So the flow of credentials would be:

Add account request to job creator:

{"account_name": "test-aws", "credentials": {"access_key_id": "0987654321", "secret_access_key": "1234567890"}, "group": "group1", "partition": "aws", "provider": "ec2", "requesting_user": "user1"}

Job creator updates the accounts file based on the account (possibly validates the credentials). It then dumps the credentials to a string and encrypts them. The message is updated and passed on to credentials:

{"account_name": "test-aws", "credentials": "encrypted string", "group": "group1", "partition": "aws", "provider": "ec2", "requesting_user": "user1"}

Credentials service just stores the encrypted string to file based on the provided account info. When a request comes in credentials service just gets the string from file and passes it in the response.

The requesting service will decrypt them back to a dictionary and use them as needed. For EC2 we would use values directly. For GCE and Azure we would temp store the credentials to use file based auth.

Hope this helps?

**Comment by rjschwei** [comment]
Lots to catch up on, a few things I noticed on first read of the discussion. I'll read it again and will probably come back with more comments.

- The end points certainly to not need to be encrypted, lets keep the bits that we encrypt ot what is necessary.
- To start with we will expect that the resource group, storage account, and container exist. As a future feature we might consider letting the user specify a command line option to create these

- It is not possible to work with temporary resource groups, storage accounts, or containers for things that get published. For images that only run to the end of testing that would be an option but then again one will probably want to run testing in a dedicated resource group. The problem is going to be when users try to run simultaneous tests in the same resource group, I think that will break things.

The uploader itself should expect that storage resource group, storage account, container exists. If we decide to create these at some point in the future we should farm that out to a utility module.

**Comment by rjschwei** [comment]
@schaefi the blob can be deleted after the disk is created

**Comment by smarlowucf** [comment]
> The end points certainly to not need to be encrypted, lets keep the bits that we encrypt ot what is necessary.

I agree they don't but if we don't encrypt them together with credentials then the credentials handling becomes very complex very quickly.

Right now it's very simple:

- All credentials will be in json format.
-The json string is encrypted and stored in file.
- The encrypted string is read and passed to services.
  - The service decrypts back to json dict and uses credentials as needed.

For azure and GCE the service account json files both contain these endpoint url's which are not secrets. But to used file based authentication we want them in credentials. And the file based authentication has very nice advantages (especially when dealing with sovereign or disconnected reegions).

So I think the advantages of sticking to a simple encryption technique outweigh the fact that we end up with some non-secret info encrypted.

**Comment by smarlowucf** [comment]
The alternative is having a complex encryption where we explicitly encrypt strings inside the dictionary and this becomes different for all providers. And end up with complexity on both ends:

if provider is azure: encrypt clientId etc etc.

{'clientId': 'encrypted_string', ..., 'activeDirectoryEndpointUrl': 'https://login.microsoftonline.com'}

if provider is azure: decrypt clientId etc.

This is certainly an option but I don't think it's worth the added complexity.

**Comment by rjschwei** [comment]
Why do we think that the endpoints need to be part of the credentials at all? I would consider those to be constants.

**Comment by smarlowucf** [comment]
> Why do we think that the endpoints need to be part of the credentials at all? I would consider those to be constants.

They are part of the service account files. https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python#mgmt-auth-file

And they are specific to each account. If the account is in a sovereign region the endpoints are different. The advantage of file base auth like this is that there's no need to handle sovereign accounts any different.

E.g. in IPA https://github.com/SUSE/ipa/blob/master/ipa/ipa_azure.py#L344. File based auth will authenticate against the correct endpoints based on the account.

**Comment by smarlowucf** [comment]
Is there a distinct disadvantage to having some non-secret info in the SA files that gets encrypted?
